File: Androlon/Sources/AndrolonEngine/Subprocess.swift
import Foundation

/// Result of running a child process. stdout and stderr are merged into
/// `output` via a single pipe — one reader means no pipe-buffer deadlock and no
/// concurrency gymnastics, which matters for chatty tools like sdkmanager.
public struct CommandResult: Sendable {
    public let status: Int32
    public let output: String
    public var ok: Bool { status == 0 }
    public var trimmed: String { output.trimmingCharacters(in: .whitespacesAndNewlines) }
}

public enum ShellError: Error, CustomStringConvertible {
    case launchFailed(tool: String, underlying: String)
    case nonZero(tool: String, result: CommandResult)

    public var description: String {
        switch self {
        case let .launchFailed(tool, underlying):
            return "failed to launch \(tool): \(underlying)"
        case let .nonZero(tool, result):
            return "\(tool) exited \(result.status):\n\(result.trimmed)"
        }
    }
}

/// Run `executable args`, blocking until exit. `extraPATH` is prepended to the
/// child PATH so bundled tools (adb/emulator) resolve; `env` overrides vars.
@discardableResult
public func runCommand(
    _ executable: URL,
    _ args: [String],
    extraPATH: [URL] = [],
    env: [String: String] = [:]
) throws -> CommandResult {
    let proc = Process()
    proc.executableURL = executable
    proc.arguments = args

    var environment = ProcessInfo.processInfo.environment
    if !extraPATH.isEmpty {
        let prefix = extraPATH.map(\.path).joined(separator: ":")
        environment["PATH"] = prefix + ":" + (environment["PATH"] ?? "")
    }
    for (key, value) in env { environment[key] = value }
    proc.environment = environment

    let pipe = Pipe()
    proc.standardOutput = pipe
    proc.standardError = pipe

    do {
        try proc.run()
    } catch {
        throw ShellError.launchFailed(tool: executable.lastPathComponent,
                                      underlying: error.localizedDescription)
    }

    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    proc.waitUntilExit()
    return CommandResult(status: proc.terminationStatus,
                         output: String(decoding: data, as: UTF8.self))
}

/// Launch a long-running process (the emulator) detached, returning it so the
/// caller can track/terminate. Output is redirected to `logFile`.
public func launchDetached(
    _ executable: URL,
    _ args: [String],
    extraPATH: [URL] = [],
    env: [String: String] = [:],
    logFile: URL
) throws -> Process {
    FileManager.default.createFile(atPath: logFile.path, contents: nil)
    guard let handle = try? FileHandle(forWritingTo: logFile) else {
        throw ShellError.launchFailed(tool: executable.lastPathComponent,
                                      underlying: "cannot open log \(logFile.path)")
    }
    let proc = Process()
    proc.executableURL = executable
    proc.arguments = args
    var environment = ProcessInfo.processInfo.environment
    if !extraPATH.isEmpty {
        let prefix = extraPATH.map(\.path).joined(separator: ":")
        environment["PATH"] = prefix + ":" + (environment["PATH"] ?? "")
    }
    for (key, value) in env { environment[key] = value }
    proc.environment = environment
    proc.standardOutput = handle
    proc.standardError = handle
    do {
        try proc.run()
    } catch {
        throw ShellError.launchFailed(tool: executable.lastPathComponent,
                                      underlying: error.localizedDescription)
    }
    return proc
}

File: crates/androlon-core/src/subprocess.rs
use crate::error::{EngineError, Result};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

/// Result of a finished child process (stdout+stderr merged).
#[derive(Debug)]
pub struct CommandOutput {
    pub status: i32,
    pub output: String,
}

impl CommandOutput {
    pub fn ok(&self) -> bool {
        self.status == 0
    }
    pub fn trimmed(&self) -> &str {
        self.output.trim()
    }
}

fn build_command(
    exe: &Path,
    args: &[&str],
    extra_path: &[PathBuf],
    env: &[(String, String)],
) -> Command {
    let mut cmd = Command::new(exe);
    cmd.args(args);
    if !extra_path.is_empty() {
        let sep = if cfg!(windows) { ';' } else { ':' };
        let prefix = extra_path
            .iter()
            .map(|p| p.display().to_string())
            .collect::<Vec<_>>()
            .join(&sep.to_string());
        let existing = std::env::var("PATH").unwrap_or_default();
        cmd.env("PATH", format!("{prefix}{sep}{existing}"));
    }
    for (k, v) in env {
        cmd.env(k, v);
    }
    cmd
}

/// Run `exe args`, blocking until exit; stdout+stderr are captured together so a
/// single reader can't deadlock on a full pipe (matters for chatty sdkmanager).
pub fn run(
    exe: &Path,
    args: &[&str],
    extra_path: &[PathBuf],
    env: &[(String, String)],
) -> Result<CommandOutput> {
    let tool = exe
        .file_name()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_else(|| exe.display().to_string());

    let out = build_command(exe, args, extra_path, env)
        .output()
        .map_err(|source| EngineError::Launch { tool: tool.clone(), source })?;

    let mut merged = String::from_utf8_lossy(&out.stdout).into_owned();
    merged.push_str(&String::from_utf8_lossy(&out.stderr));
    Ok(CommandOutput {
        status: out.status.code().unwrap_or(-1),
        output: merged,
    })
}

/// Like `run`, but errors if the process exits non-zero.
pub fn run_checked(
    exe: &Path,
    args: &[&str],
    extra_path: &[PathBuf],
    env: &[(String, String)],
) -> Result<CommandOutput> {
    let tool = exe
        .file_name()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_else(|| exe.display().to_string());
    let result = run(exe, args, extra_path, env)?;
    if result.ok() {
        Ok(result)
    } else {
        Err(EngineError::NonZero {
            tool,
            status: result.status,
            output: result.output,
        })
    }
}

/// Spawn a long-running process (the emulator) detached, redirecting output to
/// `log_file`. Returns the `Child` so the caller can track/terminate it.
pub fn spawn_detached(
    exe: &Path,
    args: &[&str],
    extra_path: &[PathBuf],
    env: &[(String, String)],
    log_file: &Path,
) -> Result<std::process::Child> {
    let tool = exe
        .file_name()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_else(|| exe.display().to_string());
    let log = std::fs::File::create(log_file)
        .map_err(|source| EngineError::Launch { tool: tool.clone(), source })?;
    let log_err = log
        .try_clone()
        .map_err(|source| EngineError::Launch { tool: tool.clone(), source })?;
    build_command(exe, args, extra_path, env)
        .stdout(Stdio::from(log))
        .stderr(Stdio::from(log_err))
        .stdin(Stdio::null())
        .spawn()
        .map_err(|source| EngineError::Launch { tool, source })
}
